app/models/concerns/impressionable.rb

```diff
@@ -1,79 +1,115 @@
+      missing_dates = dates - counts_by_date.keys
+      if missing_dates.present?
+        CreateDailySummariesJob.perform_later self, start_date.iso8601, end_date.iso8601, scoped_by
+        impressions.on(*missing_dates).scoped_by(scoped_by).group(:displayed_at_date).count
+          .each { |date, count| counts_by_date[date] ||= count }
+      end
+      dates.map { |date| counts_by_date[date] ||= 0 }
+    end
   end
 
-  def daily_impressions_count(date = nil)
-    date = Date.coerce(date)
-    key = "#{cache_key}/#{__method__}/#{date.cache_key minutes_cached: 15}"
-    Rails.cache.fetch(key) { impressions.on(date).count }.to_i
+  def total_impressions_count(start_date = nil, end_date = nil, scoped_by: nil)
+    daily_impressions_counts(start_date, end_date, scoped_by: scoped_by).sum
   end
 
-  def daily_impressions_counts(start_date = nil, end_date = nil)
-    key = "#{cache_key}/#{__method__}/#{Date.coerce(start_date).cache_key}-#{Date.coerce(end_date).cache_key}"
+  def daily_clicks_counts(start_date = nil, end_date = nil, scoped_by: nil)
+    start_date = Date.coerce(start_date)
+    end_date = Date.coerce(end_date || start_date)
+    key = "#{cache_key}/#{__method__}/#{start_date.cache_key}-#{end_date.cache_key}/#{scoped_by&.cache_key}"
     Rails.cache.fetch key do
-      results = impressions.between(start_date, end_date).group(:displayed_at_date).count
-      (start_date..end_date).map { |date| results[date] || 0 }
+      counts_by_date = daily_summaries.between(start_date, end_date).scoped_by(scoped_by)
+        .pluck(:displayed_at_date, :clicks_count)
+        .each_with_object({}) { |row, memo| memo[row[0]] = row[1] }
+      dates = (start_date..end_date).to_a
+      missing_dates = dates - counts_by_date.keys
+      if missing_dates.present?
+        CreateDailySummariesJob.perform_later self, start_date.iso8601, end_date.iso8601, scoped_by
+        impressions.clicked.on(*missing_dates).scoped_by(scoped_by).group(:displayed_at_date).count
+          .each { |date, count| counts_by_date[date] ||= count }
+      end
+      dates.map { |date| counts_by_date[date] ||= 0 }
     end
   end
 
-  def daily_impressions_per_mille(date = nil)
-    daily_impressions_count(date).to_i / 1_000.to_f
-  end
-
-  def total_clicks_count(start_date = nil, end_date = nil)
-    return daily_clicks_counts(start_date, end_date).sum if start_date && end_date
-    key = "#{cache_key}/#{__method__}/#{Date.current.cache_key minutes_cached: 15}"
-    Rails.cache.fetch(key) { impressions.clicked.count }
+  def total_clicks_count(start_date = nil, end_date = nil, scoped_by: nil)
+    daily_clicks_counts(start_date, end_date, scoped_by: scoped_by).sum
   end
 
-  def daily_clicks_count(date = nil)
-    date = Date.coerce(date)
-    key = "#{cache_key}/#{__method__}/#{date.cache_key minutes_cached: 15}"
-    Rails.cache.fetch(key) { impressions.on(date).clicked.count }.to_i
+  def total_click_rate(start_date = nil, end_date = nil, scoped_by: nil)
```

Comment:
Should this be `average_click_rate`?

Reply:
This has been renamed to simply `click_rate`

Reply:
Note: `click_rate` is not the average of daily click rates. Rather it's calculated based on impression and click counts for the given date range.